tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc

```diff
@@ -335,6 +334,161 @@ void WarnIfBadDriverJITVersion() {
+      LOG(WARNING) << "Skippping entry \"" << fullpath << "\". Can't read it.";
+      continue;
+    }
+    // store PTX or cubin in memory cache
+    in_memory_cache_[key] = text;
+  }
+  VLOG(2) << "Persistent cache has " << in_memory_cache_.size()
+          << " entries.";
+  if (VLOG_IS_ON(3)) {
+    for (const auto &pair : in_memory_cache_) {
+      VLOG(3) << "  Entry: " << pair.first << ".";
+    }
+  }
+}
+
+constexpr const int64 PersistentCompilationCache::kPtxHash;
+
+int64 PersistentCompilationCache::CreateKey(
+    llvm::Module* llvm_module, const std::pair<int, int> &compute_capability,
+    const se::GpuAsmOpts &options){
+  string llvm_str = llvm_ir::DumpModuleToString(*llvm_module);
+  std::string ptx_options;
+  if (options.disable_gpuasm_optimizations) {
+    ptx_options += "-O0";
+  }
+  for (const std::string &flag: options.extra_flags) {
+    ptx_options += flag;
+  }
+
+  int64 key = tensorflow::Hash64(llvm_str);
+  key = tensorflow::Hash64Combine(key, compute_capability.first);
+  key = tensorflow::Hash64Combine(key, compute_capability.second);
+  key = tensorflow::Hash64Combine(key, tensorflow::Hash64(ptx_options));
+
+  VLOG(3) << "Created key " << key << ".";
+ 
+  return key;
+}
+
+void PersistentCompilationCache::AddToCache(int64 key, absl::string_view text,
+                                            const string &kind) {
+  VLOG(2) << "Attempting to add " << kind << " to cache for key: "
+          << key << ".";
+  tensorflow::Env* env = tensorflow::Env::Default();
+  std::string text_tmp = tensorflow::io::JoinPath(cache_dir_, "!");
+  if (!env->CreateUniqueFileName(&text_tmp, "")) {
+    LOG(ERROR) << "Don't add to cache: cannot create a temporary file "
+               << "name to store the " << kind << ".";
+  } else {
+    if (!tensorflow::WriteStringToFile(env, text_tmp, text).ok()) {
+      LOG(ERROR) << "Don't add to cache: can't write " << kind << ". Please "
+                 << "check that there's space on the device, and that the "
+		 << "cache \"" << cache_dir_ << "\" has the right permissions.";
+    } else {
+      // Rename file.
+      string key_str = std::to_string(key);
+      string text_file = tensorflow::io::JoinPath(cache_dir_, key_str);
+      // add cache entry "key -> text".
+      // rename is atomic, making this multi thread/process safe.
+      if (!env->RenameFile(text_tmp, text_file).ok()) {
+	LOG(ERROR) << "Don't add to cache: can't rename \"" << text_tmp\
+	           << "\" to \"" << text_file << "\". Please check that "
+		   << "there's space on the device, and that the cache \""
+                   << cache_dir_ << "\" has the right permissions.";
+      } else {
+	VLOG(2) << "Added " << kind << ": " << key << " to cache directory "
+        	<< cache_dir_ << ".";
+      }
+    }
+    // On success, the temp file is no longer present, but  WriteStringToFile
+    // can fail after file has been created. Delete here in case of failures.
+    (void)env->DeleteFile(text_tmp);
```

Comment:
What if the rename succeeded, but a racing process recreated `text_tmp`?

Reply:
great catch. done.